File: OS LAB/Round_Robin.cpp
#include <iostream>
using namespace std;

int main() {

    int totalProcesses, timeQuantum;

    cout << "Enter number of processes: ";
    cin >> totalProcesses;

    cout << "Enter Time Quantum: ";
    cin >> timeQuantum;

    int processID[totalProcesses];
    int burstTime[totalProcesses];
    int remainingTime[totalProcesses];
    int completionTime[totalProcesses];
    int waitingTime[totalProcesses];

    // Input (Only Burst Time)
    for (int i = 0; i < totalProcesses; i++) {
        processID[i] = i + 1;

        cout << "Enter Burst Time for Process " << processID[i] << ": ";
        cin >> burstTime[i];

        remainingTime[i] = burstTime[i];
    }

    int currentTime = 0;
    int completed = 0;

    // Round Robin Logic
    while (completed < totalProcesses) {

        for (int i = 0; i < totalProcesses; i++) {

            if (remainingTime[i] > 0) {

                if (remainingTime[i] > timeQuantum) {
                    currentTime += timeQuantum;
                    remainingTime[i] -= timeQuantum;
                }
                else {
                    currentTime += remainingTime[i];
                    completionTime[i] = currentTime;

                    remainingTime[i] = 0;
                    completed++;
                }
            }
        }
    }

    // Calculate Waiting Time
    double totalWT = 0;

    for (int i = 0; i < totalProcesses; i++) {
        waitingTime[i] = completionTime[i] - burstTime[i];
        totalWT += waitingTime[i];
    }

    // Display
    cout << "\nPID\tBT\tCT\tWT\n";

    for (int i = 0; i < totalProcesses; i++) {
        cout << processID[i] << "\t"
             << burstTime[i] << "\t"
             << completionTime[i] << "\t"
             << waitingTime[i] << endl;
    }

    cout << "\nAverage Waiting Time: "
         << totalWT / totalProcesses << endl;

    return 0;
}

File: OS LAB/Disk Schedule Python/Round_Robin.py

n = int(input("Enter number of processes: "))

process = ["P" + str(i+1) for i in range(n)]
burst = [int(input(f"Enter Burst Time for {process[i]}: ")) for i in range(n)]
remaining = burst.copy()
waiting = [0] * n

quantum = int(input("Enter Time Quantum: "))

time = 0
while any(remaining):  
    for i in range(n):
        if remaining[i] > 0:
            if remaining[i] > quantum:
                time += quantum
                remaining[i] -= quantum
            else:
                time += remaining[i]
                waiting[i] = time - burst[i]
                remaining[i] = 0

print("\nProcess\tBurst")
for i in range(n):
    print(f"{process[i]}\t{burst[i]}")

avg_waiting = sum(waiting) / n
print("\nAverage Waiting Time =", avg_waiting)